manubot/manuscript.py

```diff
@@ -0,0 +1,64 @@
+import collections
+import pathlib
+import re
+
+from manubot.citations import citation_pattern, is_valid_citation_string
+
+
+def get_citation_strings(text):
+    """
+    Extract the deduplicated list of citations in a text
+    """
+    citations_strings = set(citation_pattern.findall(text))
+    citations_strings = filter(is_valid_citation_string, citations_strings)
+    return sorted(citations_strings)
+
+
+def get_text(directory):
+    """
+    Return a concatenated string of section texts from the specified directory.
+    """
+    section_dir = pathlib.Path(directory)
+    paths = sorted(section_dir.glob('[0-9]*.md'))
+    name_to_text = collections.OrderedDict()
+    for path in paths:
+        name_to_text[path.stem] = path.read_text()
+    return '\n\n'.join(name_to_text.values()) + '\n'
+
+
+def replace_citations_strings_with_ids(text, string_to_id):
+    """
+    Convert citations to their IDs for pandoc.
+
+    `text` is markdown source text
+
+    `string_to_id` is a dictionary like:
+    @10.7287/peerj.preprints.3100v1 → 11cb5HXoY
+    """
+    for old, new in string_to_id.items():
+        text = re.sub(
+            pattern=re.escape(old) + '(?=[^a-zA-Z0-9/])',
+            repl='@' + new,
+            string=text,
+        )
+    return text
+
+
+def get_manuscript_stats(text, citation_df):
+    """
+    Compute manuscript statistics.
+    """
+    stats = collections.OrderedDict()
+
+    # Number of distinct references by type
+    ref_counts = (
+        citation_df
+        .standard_citation
+        .drop_duplicates()
+        .map(lambda x: x.split(':')[0])
+        .pipe(collections.Counter)
+    )
+    ref_counts['total'] = sum(ref_counts.values())
+    stats['reference_counts'] = ref_counts
```

Comment:
No ATM they only get saved to `variables.json`. Hence they're available to the jinja2 templator / whenever you have access to the output files (anytime besides PR CI builds). Do you think all of variables.tsv should be logged? In what cases do you like seeing the reference and word counts?